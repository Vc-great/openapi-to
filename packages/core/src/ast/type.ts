import type { JSDocTagStructure } from "ts-morph";
import type {
  AmbientableNodeStructure,
  ExportableNodeStructure,
  JSDocableNodeStructure,
  Structure,
} from "ts-morph";
import type {
  BindingNamedNodeStructure,
  ExclamationTokenableNodeStructure,
  KindedStructure,
  OptionalKind,
  StructureKind,
  TypedNodeStructure,
  VariableDeclarationKind,
  WriterFunction,
} from "ts-morph";

export interface VariableStatementStructureOfObject
  extends Structure,
    VariableStatementSpecificStructure,
    JSDocableNodeStructure,
    AmbientableNodeStructure,
    ExportableNodeStructure {}

interface VariableStatementSpecificStructure
  extends KindedStructure<StructureKind.VariableStatement> {
  declarationKind?: VariableDeclarationKind;
  declarations: OptionalKind<VariableDeclarationStructure>[];
}

export interface VariableDeclarationStructure
  extends Structure,
    VariableDeclarationSpecificStructure,
    BindingNamedNodeStructure,
    InitializerExpressionableNodeStructure,
    TypedNodeStructure,
    ExclamationTokenableNodeStructure {}

interface VariableDeclarationSpecificStructure
  extends KindedStructure<StructureKind.VariableDeclaration> {}

export interface InitializerExpressionableNodeStructure {
  initializer?: string | WriterFunction | ObjectStructure[];
}

export interface ObjectStructure {
  key: string;
  value: string;
  docs: JSDocStructure[];
}

export interface JSDocStructure {
  /**
   * The description of the JS doc.
   * @remarks To force this to be multi-line, add a newline to the front of the string.
   */
  description?: string;
  /** JS doc tags (ex. `&#64;param value - Some description.`). */
  tags?: OptionalKind<JSDocTagStructure>[];
}

export type JSDoc = JSDocStructure[];
