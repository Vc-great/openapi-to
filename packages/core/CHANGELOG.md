# @openapi-to/core

## 3.0.1

### Patch Changes

- fix bug

## 3.0.0

### Major Changes

- fix bug
- fix: bug
- update
- refactor

## 3.0.0-alpha.5

### Major Changes

- fix: bug

## 3.0.0-alpha.4

### Major Changes

- fix bug

## 3.0.0-alpha.3

### Major Changes

- refactor

## 3.0.0-alpha.2

### Major Changes

- fix

## 2.4.0-alpha.1

### Patch Changes

- fix bug

## 2.4.0-alpha.0

### Minor Changes

- update all

## 2.3.0

### Minor Changes

- generate swr

## 2.2.0

### Minor Changes

- fix bug

## 2.1.0

### Minor Changes

- optimize request type zod

## 2.0.0

### Major Changes

- [`e519466`](https://github.com/Vc-great/openapi-to/commit/e5194667c7416e817a498d592c357a7ae9c05f22) Thanks [@Vc-great](https://github.com/Vc-great)! - v2

- [`a39a900`](https://github.com/Vc-great/openapi-to/commit/a39a9002dda434d8a65768f55c69875ed8ad1eea) - fix bug

- 2.0

- [`367be25`](https://github.com/Vc-great/openapi-to/commit/367be252aa434487c09c4566e77792839867b509) - fix bug

- [`7fd9e58`](https://github.com/Vc-great/openapi-to/commit/7fd9e58417ef5563dedf945fbb030b70b8b09bd6) - rc

## 2.0.0-rc.5

### Major Changes

- rc

## 2.0.0-alpha.4

### Major Changes

- fix bug

## 2.0.0-alpha.3

### Major Changes

- fix bug

## 2.0.0-alpha.2

### Major Changes

- v2
