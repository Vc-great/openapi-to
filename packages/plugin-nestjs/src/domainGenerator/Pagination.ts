import _ from "lodash";

import type { OpenAPIParameterObject } from "@openapi-to/core";

export class Pagination {
  constructor(
    private readonly parameters: Array<OpenAPIParameterObject> | undefined,
  ) {
    this.parameters = parameters;
  }

  get pageNoName(): string {
    return "pageNo";
  }

  get pageSizeName(): string {
    return "pageSize";
  }

  get pageNoParameterObject(): OpenAPIParameterObject | undefined {
    return _.find(this.parameters, (item) => item.name === "pageNo");
  }

  get pageSizeParameterObject(): OpenAPIParameterObject | undefined {
    return _.find(this.parameters, (item) => item.name === "pageSize");
  }

  get hasPagination(): boolean {
    const names = _.map(this.parameters, (item) => _.camelCase(item.name));

    return names.includes("pageNo") && names.includes("pageSize");
  }

  get paginationByParameters(): Array<OpenAPIParameterObject> {
    return _.chain(this.parameters)
      .filter((item: OpenAPIParameterObject) => {
        return this.hasPagination
          ? ["pageNo", "pageSize"].includes(_.camelCase(item.name))
          : false;
      })
      .value();
  }

  get filterPaginationByParameters(): Array<OpenAPIParameterObject> {
    return _.chain(this.parameters)
      .filter((item: OpenAPIParameterObject) => {
        return this.hasPagination
          ? !["pageNo", "pageSize"].includes(_.camelCase(item.name))
          : true;
      })
      .value();
  }

  get addPaginationDtoName() {
    return this.hasPagination ? "PaginationDto" : undefined;
  }
}
