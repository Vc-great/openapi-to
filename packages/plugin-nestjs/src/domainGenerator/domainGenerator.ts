import _ from "lodash";
import { StructureKind } from "ts-morph";

import { NestjsGenerator } from "../NestjsGenerator.ts";
import { Component } from "./Component.ts";
import { Pagination } from "./Pagination.ts";
import { Schema } from "./Schema.ts";
import { SwaggerGenerator } from "./SwaggerGenerator.ts";

import type {
  OpenAPIParameterObject,
  OpenAPIRequestBodyObject,
  OpenAPIResponseObject,
  PluginContext,
} from "@openapi-to/core";
import type { SchemaObject } from "oas/types";
import type { OptionalKind, PropertyDeclarationStructure } from "ts-morph";
import type { ClassDeclarationStructure } from "ts-morph";
import type { Config } from "../types.ts";

export class DomainGenerator extends NestjsGenerator {
  private component: Component;
  private schema: Schema;

  private swaggerGenerator: SwaggerGenerator;

  constructor(config: Config) {
    super(config);

    this.schema = new Schema(config);
    this.component = new Component(config);

    this.swaggerGenerator = new SwaggerGenerator(config);
  }

  build(context: PluginContext): void {
    _.forEach(this.openapi.pathGroupByTag, (pathGroup, tag) => {
      _.forEach(pathGroup, ({ path, method, tag }) => {
        this.operation = this.openapi.setCurrentOperation(path, method, tag);

        this.generateParametersDto();
        this.generateRequestBodyDto();
        this.generateResponseDto();
      });
    });
    this.component.build();
  }

  generateQueryDto() {
    const pagination = new Pagination(
      this.openapi.parameter?.parametersOfQuery,
    );
    const parameters = pagination.filterPaginationByParameters;

    const queryExtends = _.chain(parameters)
      .filter((item: OpenAPIParameterObject) => Boolean(item.$ref))
      .map((item) => _.upperFirst(item.refName as string))
      .concat(pagination.addPaginationDtoName ?? [])
      .join(",")
      .value();

    const queryStatements: Omit<ClassDeclarationStructure, "kind"> = {
      leadingTrivia: "\n",
      isExported: true,
      extends: queryExtends ? `IntersectionType(${queryExtends})` : "",
      name: this.queryDtoName,
      decorators: [],
      docs: [],
      methods: [],
      properties: _.chain(this.openapi.parameter?.parametersOfQuery)
        .filter((item) => !item.$ref)
        .map((item) => {
          const schema = item.schema as SchemaObject;
          const validatorDecorator =
            this.validatorDecorator.generatorDecorator(schema, item.required) ??
            [];
          return {
            leadingTrivia: "\n",
            kind: StructureKind.PropertySignature,
            name: item.name,
            type: schema ? this.schema.getSchemaType(schema) : "unknown",
            hasQuestionToken: !item.required,
            initializer: JSON.stringify(schema?.default) || "",
            decorators: [
              ...validatorDecorator,
              ...(item
                ? this.swaggerGenerator.generatorDecoratorFromParameterObject(
                    item,
                  )
                : []),
            ].filter(Boolean),
          };
        })
        //todo type
        .value() as OptionalKind<PropertyDeclarationStructure>[],
    };

    if (!_.isEmpty(queryStatements.properties)) {
      const parametersOfQuery = this.openapi.parameter?.parametersOfQuery;
      this.ast.createSourceFile(this.queryDtoFilePath, {
        statements: [
          ...this.ast.generateImportStatements(
            this.mergeNamedImports(
              [queryExtends ? this.intersectionTypeImport : undefined]
                .concat(
                  parametersOfQuery
                    ? [
                        this.validatorImportsOfParameters(parametersOfQuery),
                        this.swaggerGenerator.apiPropertyImport,
                      ]
                    : undefined,
                )
                .filter(Boolean),
            ),
          ),
          this.ast.generateClassStatements(queryStatements),
        ],
      });
    }
  }

  generateParametersDto(): void {
    this.generateQueryDto();
  }

  generateResponseDto(): void {
    this.generateResponseSingleSchema(this.openapi.response?.successResponse);
  }

  generateResponseSingleSchema(
    responseObject: OpenAPIResponseObject | undefined,
  ): void {
    if (responseObject === undefined) {
      return;
    }
    if (!responseObject.refName && !responseObject.schema) {
      return;
    }

    const schema = responseObject?.schema;
    const description = responseObject?.description;

    const className = `${_.upperFirst(this.operation?.getOperationId())}Vo`;

    const statements: Omit<ClassDeclarationStructure, "kind"> = {
      leadingTrivia: "\n",
      isExported: true,
      extends: responseObject.refName
        ? `IntersectionType(${responseObject.refName})`
        : undefined,
      name: `${className}`,
      decorators: [],
      docs: description
        ? [
            {
              description: "\n",
              tags: [
                {
                  tagName: "description",
                  text: description,
                },
              ],
            },
          ]
        : [],
      methods: [],
      properties: responseObject.refName
        ? []
        : (this.schema.getPropertyStructureFromSchemaObject(schema)
            ?.propertiesStructure ?? []),
    };

    this.ast.createSourceFile(this.responseDtoFilePath, {
      statements: [
        ...this.ast.generateImportStatements(
          this.mergeNamedImports(
            [
              this.generateCommonDomainImports(
                responseObject.refName ? [responseObject.refName] : undefined,
              ),
            ]
              .concat(
                _.isEmpty(statements.properties)
                  ? []
                  : this.swaggerGenerator.apiPropertyImport,
              )
              .concat(
                responseObject.refName
                  ? [this.intersectionTypeImport]
                  : [this.validatorImportsOfSchema(schema)],
              )
              .filter(Boolean),
          ),
        ),
        this.ast.generateClassStatements(statements),
      ],
    });
  }

  generateRequestBodyDto(): void {
    const requestBodyObject = this.openapi.requestBody
      ?.requestBodyObject as OpenAPIRequestBodyObject;

    if (requestBodyObject === undefined) {
      return;
    }

    if (!requestBodyObject.refName && !requestBodyObject.schemaObject) {
      return;
    }

    const className: string =
      requestBodyObject.refName ||
      this.openapi.getDomainNameByRef(
        _.get(requestBodyObject, "schemaObject.$ref", ""),
      ) ||
      `${this.operation?.getOperationId()}Dto`;

    const schema =
      requestBodyObject.schemaObject &&
      "$ref" in requestBodyObject.schemaObject &&
      requestBodyObject.schemaObject.$ref
        ? (this.openapi.findSchemaDefinition(
            requestBodyObject?.schemaObject.$ref,
          ) as SchemaObject)
        : (requestBodyObject.schemaObject as SchemaObject);

    const propertyStructures =
      this.schema.getPropertyStructureFromSchemaObject(schema);
    const statements: Omit<ClassDeclarationStructure, "kind"> = {
      leadingTrivia: "\n",
      isExported: true,
      extends: undefined,
      name: `${className}`,
      decorators: [],
      docs: requestBodyObject.description
        ? [
            {
              description: "\n",
              tags: [
                {
                  tagName: "description",
                  text: requestBodyObject.description,
                },
              ],
            },
          ]
        : [],
      methods: [],
      properties: _.isEmpty(propertyStructures?.propertiesStructure)
        ? []
        : propertyStructures?.propertiesStructure,
    };

    if (!_.isEmpty(statements.properties)) {
      this.ast.createSourceFile(this.bodyDtoFilePath, {
        statements: [
          ...this.ast.generateImportStatements(
            this.mergeNamedImports(
              [this.validatorImportsOfSchema(schema)]
                .concat(
                  _.isEmpty(propertyStructures?.propertiesStructure)
                    ? []
                    : this.swaggerGenerator.apiPropertyImport,
                )
                .concat(
                  schema ? this.generateDomainImportsBySchema(schema) : [],
                )
                .filter(Boolean),
            ),
          ),
          this.ast.generateClassStatements(statements),
        ],
      });
    }
  }
}
