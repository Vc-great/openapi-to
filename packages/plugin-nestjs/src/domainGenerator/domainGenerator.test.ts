import path from "node:path";
import { dirname } from "node:path";
import { fileURLToPath } from "node:url";

import { AST, OpenAPI } from "@openapi-to/core";

import _ from "lodash";
import Oas from "oas";
import { describe } from "vitest";

import petstore from "../../mock/petstore.json";
import usermock from "../../mock/user.json";
import { DomainGenerator } from "./domainGenerator.ts";

import type { OpenapiToSingleConfig } from "@openapi-to/core";
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe("domainGenerator", async () => {
  const openapiToSingleConfig: OpenapiToSingleConfig = {
    name: "",
    input: {
      path: ""
    },
    output: {
      dir: path.resolve(__dirname, "../mock"),
    },
    plugins: [],
  };

  test("domainGenerator", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    // const oas = new Oas(petStore);
    const oas = new Oas(usermock);
    const openapi = new OpenAPI({}, oas);
    const pluginConfig = {};
    const context = {
      output: "",
    };
    const domainGenerator = new DomainGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    domainGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });

  test("domainGenerator petstore", () => {
    const ast = new AST();
    // @ts-expect-error Not a canonical document
    // const oas = new Oas(petStore);
    const oas = new Oas(petstore);
    const openapi = new OpenAPI({}, oas);
    const pluginConfig = {};
    const context = {
      output: "",
    };
    const domainGenerator = new DomainGenerator({
      oas,
      ast,
      openapi,
      pluginConfig,
      openapiToSingleConfig,
    });
    domainGenerator.build(context);
    const text = _.chain(ast.sourceFile)
      .map((sourceFile) => sourceFile.getFullText())
      .join("\n")
      .value();

    expect(text).toMatchSnapshot();
  });
});
