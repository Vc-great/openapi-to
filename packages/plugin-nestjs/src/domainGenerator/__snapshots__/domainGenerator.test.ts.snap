// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`domainGenerator > domainGenerator 1`] = `
"import { User } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description OK
 */
export class Get-userVo extends IntersectionType(User) {
}

import { IntersectionType, ApiProperty } from "@nestjs/swagger";
import { IsString, Type, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class FindAllQueryDto extends IntersectionType(Ref1, PaginationDto) {

    @IsString()
    @ApiProperty({
        description: '创建时间',
        format: 'date',
        required: true
    })
    createTime: string;

    @Type(() => String)
    @IsString({ "each": true })
    @IsOptional()
    @ApiProperty({
        description: '时间',
        example: [
            '1',
            '2'
        ],
        uniqueItems: true,
        isArray: true
    })
    time?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: 'a:1,b,c:3',
        enum: [
            'a',
            'b',
            'c'
        ]
    })
    enumString?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: '枚举值',
        enum: [
            'a',
            'b',
            'c'
        ],
        default: 'a'
    })
    enum1?: string = "a";
}

import { User } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description 我是FindOne的描述
 */
export class FindAllVo extends IntersectionType(User) {
}

import { CreateUserDTO } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description OK
 */
export class Post-petVo extends IntersectionType(CreateUserDTO) {
}

import { Type, IsString, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";
import { order } from "order";


/**
 *
 * @description 我是bodyschema描述
 */
export class put-petDto {

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'a'
    })
    a ?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'b'
    })
    b ?: string;

    @IsOptional()
    @ApiProperty({
        $ref: '#/components/schemas/order',
        required: false,
        name: 'c'
    })
    c ?: order;

    @Type(() => Object)
    @IsOptional()
    @ApiProperty({
        isArray: true,
        required: false,
        name: 'data'
    })
    data ?: object[];
}

import { IsString, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class Ref1 {

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: 'ref 引入',
        required: false
    })
    ref?: string;
}

import { IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class Times {

    @IsOptional()
    @ApiProperty({
        description: '时间啊',
        format: 'time',
        required: false
    })
    times?: unknown;
}

import { Type, IsString, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class Sort {

    @Type(() => String)
    @IsString({ "each": true })
    @IsOptional()
    @ApiProperty({
        description: '排序',
        example: '[\\'id:asc\\',\\'name:desc\\']',
        isArray: true
    })
    sort?: string[];
}

import { IsNumber } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class Id {

    @IsNumber()
    @ApiProperty({
        description: 'path id',
        required: true
    })
    id: number;
}

import { IsString, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class UserId {

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: '用户id',
        required: false
    })
    userId?: string;
}

import { IsString, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class Cooickid {

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: 'cooickid',
        required: false
    })
    cooickid?: string;
}

import { IsNumber, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class PaginationDto {

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        description: '当前页',
        exclusiveMinimum: true,
        minimum: 1,
        default: 1
    })
    pageNo?: number = 1;

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        description: '数据条数',
        format: 'float'
    })
    pageSize?: number;
}
"
`;

exports[`domainGenerator > domainGenerator petstore 1`] = `
"import { IsString, IsOptional, IsBoolean, Type, IsNumber } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class FindAllQueryDto {

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false
    })
    email?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false
    })
    name?: string;

    @IsBoolean()
    @IsOptional()
    @ApiProperty({
        required: false
    })
    bol?: boolean;

    @Type(() => String)
    @IsString({ "each": true })
    @IsOptional()
    @ApiProperty({
        description: '排序,默认排序顺序为升序',
        isArray: true,
        required: false
    })
    sort?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false
    })
    createTime?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: '0:第一个枚举值,1:第二个枚举值,3:第三个枚举值',
        default: '123',
        enum: [
            '1',
            '2',
            '3'
        ],
        required: false
    })
    status?: string = "123";

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        description: '页码 (0..N)',
        default: 1,
        required: false
    })
    page?: number = 1;

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        description: '每页显示条目个数',
        default: 10,
        required: false
    })
    pageSize?: number = 10;
}

import { FindAllUserVo } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description success111111
 */
export class FindAllVo extends IntersectionType(FindAllUserVo) {
}

import { Type, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";
import { test321 } from "test321";


/**
 *
 * @description resources
 */
export class testBodyDto {

    @Type(() => test321)
    @IsOptional()
    @ApiProperty({
        description: 'test321',
        isArray: true,
        required: false,
        name: 'test321'
    })
    test321?: test321;
}

import { testDto2 } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description OK
 */
export class TestidgetVo extends IntersectionType(testDto2) {
}

import { Type, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";
import { test321 } from "test321";


/**
 *
 * @description resources
 */
export class testBodyDto {

    @Type(() => test321)
    @IsOptional()
    @ApiProperty({
        description: 'test321',
        isArray: true,
        required: false,
        name: 'test321'
    })
    test321?: test321;
}

import { testDto2 } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description OK
 */
export class TestidgetVo extends IntersectionType(testDto2) {
}

import { test32145 } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description OK
 */
export class TestdeleteVo extends IntersectionType(test32145) {
}

import { testDto2 } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description OK
 */
export class TestidgetVo extends IntersectionType(testDto2) {
}

import { IntersectionType, ApiProperty } from "@nestjs/swagger";
import { Type, IsString, IsOptional, IsBoolean } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class TestidgetQueryDto extends IntersectionType(Sort, PaginationDto) {

    @Type(() => String)
    @IsString({ "each": true })
    @IsOptional()
    @ApiProperty({
        description: 'fields',
        isArray: true,
        required: false
    })
    fields?: string;

    @IsString()
    @ApiProperty({
        format: 'date',
        default: '2024-01-01 11:11:11',
        required: true
    })
    createTime: string = "2024-01-01 11:11:11";

    @Type(() => String)
    @IsString({ "each": true })
    @IsOptional()
    @ApiProperty({
        example: [
            '2022-01-02 11:11:11',
            '2022-01-02 11:12:11'
        ],
        uniqueItems: true,
        isArray: true
    })
    time?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: 'a:1,b,c:3',
        default: 'a',
        enum: [
            'a',
            'b',
            'c'
        ]
    })
    enumString?: string = "a";

    @IsBoolean()
    @IsOptional()
    @ApiProperty({
        description: '布尔值',
        default: false
    })
    bool?: boolean = false;
}

import { testDto2 } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description OK
 */
export class TestidgetVo extends IntersectionType(testDto2) {
}

import { ApiResponse } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description successful operation
 */
export class UploadFileVo extends IntersectionType(ApiResponse) {
}

import { Type, IsNumber, IsOptional, IsString } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";
import { Category } from "Category";
import { Tag } from "Tag";


/**
 *
 * @description Pet object that needs to be added to the store
 */
export class Pet {

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        format: 'int64',
        required: false,
        name: 'id'
    })
    id?: number;

    @IsOptional()
    @ApiProperty({
        $ref: '#/components/schemas/Category',
        required: false,
        name: 'category'
    })
    category?: Category;

    @IsString()
    @ApiProperty({
        example: 'doggie',
        required: true,
        name: 'name'
    })
    name: string;

    @Type(() => String)
    @IsString({ "each": true })
    @ApiProperty({
        xml: {
            wrapped: true
        },
        isArray: true,
        required: true,
        name: 'photoUrls'
    })
    photoUrls: string[];

    @Type(() => Tag)
    @IsOptional()
    @ApiProperty({
        xml: {
            wrapped: true
        },
        isArray: true,
        required: false,
        name: 'tags'
    })
    tags?: Tag;

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: 'pet status in the store',
        enum: [
            'available',
            'pending',
            'sold'
        ],
        required: false,
        name: 'status'
    })
    status?: string;
}

import { Type, IsNumber, IsOptional, IsString } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";
import { Category } from "Category";
import { Tag } from "Tag";


/**
 *
 * @description Pet object that needs to be added to the store
 */
export class Pet {

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        format: 'int64',
        required: false,
        name: 'id'
    })
    id?: number;

    @IsOptional()
    @ApiProperty({
        $ref: '#/components/schemas/Category',
        required: false,
        name: 'category'
    })
    category?: Category;

    @IsString()
    @ApiProperty({
        example: 'doggie',
        required: true,
        name: 'name'
    })
    name: string;

    @Type(() => String)
    @IsString({ "each": true })
    @ApiProperty({
        xml: {
            wrapped: true
        },
        isArray: true,
        required: true,
        name: 'photoUrls'
    })
    photoUrls: string[];

    @Type(() => Tag)
    @IsOptional()
    @ApiProperty({
        xml: {
            wrapped: true
        },
        isArray: true,
        required: false,
        name: 'tags'
    })
    tags?: Tag;

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: 'pet status in the store',
        enum: [
            'available',
            'pending',
            'sold'
        ],
        required: false,
        name: 'status'
    })
    status?: string;
}

import { Type, IsString } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class FindPetsByStatusQueryDto {

    @Type(() => String)
    @IsString({ "each": true })
    @ApiProperty({
        description: 'Status values that need to be considered for filter',
        isArray: true,
        required: true
    })
    status: string;
}

import { Pet } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description successful operation
 */
export class GetPetByIdVo extends IntersectionType(Pet) {
}

import { Type, IsString } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class FindPetsByTagsQueryDto {

    @Type(() => String)
    @IsString({ "each": true })
    @ApiProperty({
        description: 'Tags to filter by',
        isArray: true,
        required: true
    })
    tags: string;
}

import { Pet } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description successful operation
 */
export class GetPetByIdVo extends IntersectionType(Pet) {
}

import { Pet } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description successful operation
 */
export class GetPetByIdVo extends IntersectionType(Pet) {
}

import { Type, IsNumber, IsOptional, IsString, IsBoolean } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


/**
 *
 * @description order placed for purchasing the pet
 */
export class Order {

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        format: 'int64',
        required: false,
        name: 'id'
    })
    id?: number;

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        format: 'int64',
        required: false,
        name: 'petId'
    })
    petId?: number;

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        format: 'int32',
        required: false,
        name: 'quantity'
    })
    quantity?: number;

    @IsString()
    @IsOptional()
    @ApiProperty({
        format: 'date-time',
        required: false,
        name: 'shipDate'
    })
    shipDate?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        description: 'Order Status',
        enum: [
            'placed',
            'approved',
            'delivered'
        ],
        required: false,
        name: 'status'
    })
    status?: string;

    @IsBoolean()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'complete'
    })
    complete?: boolean;
}

import { Order } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description successful operation
 */
export class GetOrderByIdVo extends IntersectionType(Order) {
}

import { Order } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description successful operation
 */
export class GetOrderByIdVo extends IntersectionType(Order) {
}

import { Type, IsOptional } from "class-validator";


/**
 *
 * @description successful operation
 */
export class GetInventoryVo {
}

import { User } from "@/common/domain";
import { IntersectionType } from "@nestjs/swagger";


/**
 *
 * @description successful operation
 */
export class GetUserByNameVo extends IntersectionType(User) {
}

import { Type, IsNumber, IsOptional, IsString } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


/**
 *
 * @description Created user object
 */
export class User {

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        format: 'int64',
        required: false,
        name: 'id'
    })
    id?: number;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'username'
    })
    username?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'firstName'
    })
    firstName?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'lastName'
    })
    lastName?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'email'
    })
    email?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'password'
    })
    password?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'phone'
    })
    phone?: string;

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        description: 'User Status',
        format: 'int32',
        required: false,
        name: 'userStatus'
    })
    userStatus?: number;
}

import { IsString } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class LoginUserQueryDto {

    @IsString()
    @ApiProperty({
        description: 'The user name for login',
        required: true
    })
    username: string;

    @IsString()
    @ApiProperty({
        description: 'The password for login in clear text',
        required: true
    })
    password: string;
}

import { IsString, IsOptional } from "class-validator";


/**
 *
 * @description successful operation
 */
export class LoginUserVo {
}

import { Type, IsNumber, IsOptional, IsString } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


/**
 *
 * @description Created user object
 */
export class User {

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        format: 'int64',
        required: false,
        name: 'id'
    })
    id?: number;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'username'
    })
    username?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'firstName'
    })
    firstName?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'lastName'
    })
    lastName?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'email'
    })
    email?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'password'
    })
    password?: string;

    @IsString()
    @IsOptional()
    @ApiProperty({
        required: false,
        name: 'phone'
    })
    phone?: string;

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        description: 'User Status',
        format: 'int32',
        required: false,
        name: 'userStatus'
    })
    userStatus?: number;
}

import { IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class Sort {

    @IsOptional()
    @ApiProperty({
        description: '排序',
        example: '[\\'id:asc\\',\\'name:desc\\']',
        isArray: true
    })
    sort?: any[];
}

import { IsNumber, IsOptional } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";


export class PaginationDto {

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        description: '当前页',
        exclusiveMinimum: true,
        minimum: 1,
        default: 1
    })
    pageNo?: number = 1;

    @IsNumber()
    @IsOptional()
    @ApiProperty({
        description: '数据条数',
        format: 'float',
        default: 10
    })
    pageSize?: number = 10;
}
"
`;
